### PR TITLE
ENH: add validation split

### PR DESCRIPTION
**Pull request by pierreablin**
This PR adds a validation split to the train / val / test. The test data should only be used to report final perf, validation set is to be used to early stop.

### CODE REVIEW

**Comment by pierreablin** [comment]
This is a WIP

File: utils/multi_frameworks_dataset.py [hunk 1]
```diff
@@ -78,23 +94,32 @@ def get_tf_data(self):
...
+            "test",
+        ]
+        for key, split in zip(["dataset", "val_dataset", "test_dataset"], splits):
             ds = tfds.load(
                 self.tf_ds_name,
                 split=split,
                 as_supervised=True,
             )
-            if key == 'test_dataset':
+            if key == "test_dataset":
```
Comment (zaccharieramzi) [inline utils/multi_frameworks_dataset.py:108]:
```suggestion
            if key != "dataset":
```

File: datasets/simulated.py [hunk 2]
```diff
@@ -41,79 +41,119 @@ def __init__(
         img_size=32,
         n_classes=2,
         train_frac=0.8,
-        framework='pytorch',
+        val_frac=0.8,
```
Comment (tomMoral) [inline datasets/simulated.py:44]: This name is inconsistent, should be `0.2` and using `1-val_frac` for `n_train` no?

Comment (zaccharieramzi) [inline datasets/simulated.py:44]: I would rename `train_frac`, as `train_val_frac` and `val_frac` to `train_frac`, in order to have a clearer picture

**Review by tomMoral — COMMENTED** [review]
My main concern is that with this, `train` and `valid` loss are not really comparable between `torch` and `tf` as tey do not rely on the same split...
Not sure if there is a way to fix this however as the order of the samples are not the same in the initial dataset.

**Review by zaccharieramzi — COMMENTED** [review]
I know this is WIP but I already have a few comments
Also don't forget to update the ds description for svhn

**Comment by pierreablin** [comment]
@tomMoral yes this is what I'll try to do next.

**Comment by tomMoral** [comment]
An idea:

why not compute a matching first between the datasets in `torch` and `tf`, store it a file (potentially on OSF but I am fine in the repo to start with), and fix this order when loading the data?

that way we can do the `train/valid` split in the `objective`. This also solves the issue with the slow data consistency as the order should always be the same, so no need to recompute it anymore.

WDYT @zaccharieramzi ?

**Comment by zaccharieramzi** [comment]
I think it's a good idea

File: build_dataset_correspondance_torch_tf.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        for train in [True, False]:
+            print(f"Registration for {dataset}, train = {train}")
+            X1, y1 = get_numpy_from_torch(dataset, train)
+            X2, y2 = get_numpy_from_tf(dataset, train)
+            perm = find_permutation_labels(X1, y1, X2, y2)
+            trainstr = "train" if train else "test"
+            np.save(
+                f"./torch_tf_datasets_registrations/{dataset}_{trainstr}.npy",
+                perm,
+            )
```
Comment (tomMoral) [inline build_dataset_correspondance_torch_tf.py:84]: This looks super nice!! thx @pierreablin 

Now we just need to use them to reorder the samples?

**Comment by zaccharieramzi** [comment]
All green in local

File: utils/multi_frameworks_dataset.py [hunk 4]
```diff
@@ -19,8 +30,32 @@ class MultiFrameworkDataset(BaseDataset, ABC):
...
+            np.arange(len(registration_indices)),
+            test_size=self.ds_description["n_samples_val"],
+            train_size=self.ds_description["n_samples_train"],
+            random_state=self.random_state,
+        )
+        self.tf_train_indices = np.sort(self.tf_train_indices)
+        self.tf_val_indices = np.sort(self.tf_val_indices)
+        self.torch_train_indices = registration_indices[self.tf_train_indices]
+        self.torch_val_indices = registration_indices[self.tf_val_indices]
+        self.train_val_split_spec = True
```
Comment (tomMoral) [inline utils/multi_frameworks_dataset.py:58]: Not a big fan of storing all this and pass it in an implicit manners.
I would do something like:
```suggestion
        registration_indices = self.get_registration_indices()

        tf_train_indices, tf_val_indices = train_test_split(
            np.arange(len(registration_indices)),
            test_size=self.ds_description["n_samples_val"],
            train_size=self.ds_description["n_samples_train"],
            random_state=self.random_state,
        )
        tf_train_indices.sort(), tf_val_indices.sort()
        if self.framework == "tensorflow" or registration_indices is None:
            return tf_train_indices, tf_val_indices

        torch_train_indices = registration_indices[tf_train_indices]
        torch_val_indices = registration_indices[tf_val_indices]
        return torch_train_indices, torch_val_indices
```

File: utils/multi_frameworks_dataset.py [hunk 5]
```diff
@@ -48,19 +83,39 @@ def get_torch_data(self):
...
+            val_dataset = Subset(data_dict["dataset"], self.torch_val_indices)
+        else:
+            train_dataset, val_dataset = random_split(
+                data_dict["dataset"],
+                [
+                    self.ds_description["n_samples_train"],
+                    self.ds_description["n_samples_val"],
+                ],
+                generator=torch.Generator().manual_seed(self.random_state),
+            )
```
Comment (tomMoral) [inline utils/multi_frameworks_dataset.py:112]:
```suggestion
        train_idx, val_idx = self.get_train_val_indices()
        train_dataset = Subset(data_dict["dataset"], train_idx)
        val_dataset = Subset(data_dict["dataset"], val_idx)
```

File: utils/multi_frameworks_dataset.py [hunk 6]
```diff
@@ -81,23 +135,54 @@ def get_tf_data(self):
...
+                    image_preprocessing(x),
+                    y,
+                ),
+                num_parallel_calls=tf.data.experimental.AUTOTUNE,
+            )
+            data_dict["dataset"] = filter_ds_on_indices(
+                data_dict["dataset"],
+                self.tf_train_indices,
+            )
+        return "object", data_dict
```
Comment (tomMoral) [inline utils/multi_frameworks_dataset.py:179]: With the proposed changes.
```suggestion
        
        train_idx, val_idx = self.get_train_val_indices()
        data_dict["val_dataset"] = filter_ds_on_indices(
            data_dict["dataset"], val_idx,
        ).map(
            lambda x, y: (
                image_preprocessing(x),
                y,
            ),
            num_parallel_calls=tf.data.experimental.AUTOTUNE,
        )
        data_dict["dataset"] = filter_ds_on_indices(
            data_dict["dataset"], train_idx
        )
        return data_dict
```

File: utils/multi_frameworks_dataset.py [hunk 7]
```diff
@@ -81,23 +135,54 @@ def get_tf_data(self):
...
 
     def get_data(self):
         """Switch to select the data from the right framework."""
+        self.random_state = 42  # Hackish
+        self.set_train_val_indices()
```
Comment (tomMoral) [inline utils/multi_frameworks_dataset.py:184]:
```suggestion
```

**Review by tomMoral — APPROVED** [review]
LGTM! thank you both!

Just a small proposal to always split with `sklearn` to make the code a bit less complex?

File: build_dataset_correspondance_torch_tf.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    for dataset in datasets:
+        for train in [True, False]:
+            print(f"Registration for {dataset}, train = {train}")
+            trainstr = "train" if train else "test"
+            registration_dir = Path("./torch_tf_datasets_registrations/")
+            filepath = registration_dir / f"{dataset}_{trainstr}.npy"
+            if args.force_registration or not filepath.exists():
+                X_torch, y_torch = get_numpy_from_torch(dataset, train)
+                X_tf, y_tf = get_numpy_from_tf(dataset, train)
+                perm = find_permutation_labels(X_torch, y_torch, X_tf, y_tf)
```
Comment (zaccharieramzi) [inline build_dataset_correspondance_torch_tf.py:88]:
```suggestion
                perm = find_permutation_per_labels(X_torch, y_torch, X_tf, y_tf)
```

File: tests/test_dataset_consistency.py [hunk 9]
```diff
@@ -137,78 +108,32 @@ def test_datasets_consistency(dataset_module_name, dataset_type):
...
-                matched_torch_indices,
-                unmatched_tf_indices,
-                unmatched_torch_indices,
-            ) = get_matched_unmatched_indices_arrays(
-                X_tf,
-                X_torch_channel_last,
-                stride
+    if dataset_type == 'test_dataset' and dataset_module_name != 'simulated':
+        registration_dir = Path("./torch_tf_datasets_registrations/")
+        tf_ds_name = {
```
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply tests/test_dataset_consistency.py]: you're right

File: objective.py [hunk 10]
```diff
@@ -61,6 +61,7 @@ class Objective(BaseObjective):
         'pip:torch', 'pip:torchvision', 'pip:pytorch-lightning ',
         # TODO: rm below, and fix tests
         'pip:tensorflow-datasets', 'pip:tensorflow-addons',
+        "scikit-learn",
```
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply objective.py:64]: it's the bug we had here: https://github.com/benchopt/benchmark_resnet_classif/pull/26#pullrequestreview-961571816 (still unexplained)

File: utils/multi_frameworks_dataset.py [hunk 11]
```diff
@@ -19,8 +30,32 @@ class MultiFrameworkDataset(BaseDataset, ABC):
...
+        filepath = registration_dir / f"{self.tf_ds_name}_train.npy"
+        if filepath.exists():
+            return np.load(filepath)
+        print('Registration file not found')
+        return None
+
+    def set_train_val_indices(self):
+        registration_indices = self.get_registration_indices()
+        if registration_indices is None:
+            self.train_val_split_spec = False
```
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply utils/multi_frameworks_dataset.py:47]: yes !

[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply utils/multi_frameworks_dataset.py:47]:
```suggestion
            self.train_val_split_spec = False
            return None
```

**Review by tomMoral — APPROVED** [review]
LGTM!

**Comment by zaccharieramzi** [comment]
all green in local, just this problem of the CIFAR DL again, submitting a tentative PR to fix the robustness

**Comment by zaccharieramzi** [comment]
no time to wait for the macos workers
